enhancements/support-cluster-HA-API.md

```diff
@@ -0,0 +1,134 @@
+
+The `ClusterConfig` struct can be defined to be:
+
+```go
+import (
+    ...
+    apiv1 "github.com/openshift/api/config/v1"
+    ...
+)
+
+// ClusterConfig contains the configuration information of the
+// cluster.
+type ClusterConfig struct {
+    // ControlPlaneTopology contains the expectations for operands that 
+    // run on control plane nodes.
+    ControlPlaneTopology apiv1.TopologyMode
+
+    // InfrastructureTopology contains the expectations for infrastructre
+    // services.
+    InfrastructureTopology apiv1.TopologyMode
+}
+```
+
+The `ClusterConfig` contains the high availability expectations for control plane and infrastructure nodes. The `apiv1.TopologyMode` is defined by OpenShift API and refers to the [topology mode][api_topology_mode_pr] in which the cluster is running. It can have the following two values:
+
+1. `ControlPlaneTopology` - refers to the expectations for operands that normally run on [control plane][control_plane] (or master) nodes.
+2. `InfrastructureTopology` - refers to the expectations for infrastructure services that do not run on control plane nodes, usually indicated by a node selector for a `role` value other than `master`.
+
+The possible values for both can either be `HighlyAvailable` or `SingleReplica`. More details on when the cluster is considered to support highly available or single replica deployments can be found [here][enhancement_cluster_operators].
+
+**Note**
+In future, the `ClusterConfig` can be extended to store additional information on cluster topology. 
+
+The helper function will return the `ClusterConfig` mentioned above by reading the `infrastructure.config.openshift.io` crd present in cluster. The user will be expected to provide the location of the `kubeconfig` file, with which an existing  cluster can be accessed using [Go client][upstream_go_client].
+
+```go
+// ErrAPINotAvailable is the error returned when the infrastructure CRD
+// is not found in cluster. This means that this could be a non-openshift cluster.
+var ErrAPINotAvailable = fmt.Errorf("infrastructure crd cannot be found in cluster")
+
+// GetClusterConfig will return the HA expectations of the cluster
+func GetClusterConfig(kubeconfigPath string) (ClusterConfig, error) {
+    // Implementation will involve reading the required fields
+    // in status subresource of `infrastructure.config.openshift.io`
+    // crd. If the crd is not present this will return an 
+    // `ErrAPINotAvailable` error.
+}
+```
+
+**NOTE:**
+Operators cannot modify the informational resources present in OpenShift clusters. The function can only be used to retrieve information about the high availability expectations of the cluster and not to modify it.
+
+### Scenarios:
+
+1. Operator author tries to fetch the `ClusterConfig` for a non-openshift cluster:
+
+In this case, an `ErrAPINotAvailable` error is returned. The error can be wrapped, such that users can utilize the `errors.Is` method to identify this specific case.
```

Comment:
When the Infrastructure resource is not present, why return an error instead of default or nil values?

Reply:
Isn't it better to explicitly state out that this particular feature is for openshift clusters only? Users can chose to handle the error gracefully and proceed if they use this in a non-openshift cluster, since this entire feature is dependent on a specific CRD which is present only in openshift. Would it be fine to say that we have a HA/non-HA detector for the cluster which does not expose such a detail itself.

Reply:
> Isn't it better to explicitly state out that this particular feature is for openshift clusters only? 
> Users can chose to handle the error gracefully and proceed if they use this in a non-openshift cluster, since this entire feature is dependent on a specific CRD which is present only in openshift. 

I would expect an helper function for an OpenShift cluster to return the Infrastructure resource directly. But that helper wouldn't be all that useful, since it's just the Get() method of the client.

> Would it be fine to say that we have a HA/non-HA detector for the cluster which does not expose such a detail itself.

I'm not sure what you mean here. Can you give an example of what that might look like?

Reply:
> I would expect an helper function for an OpenShift cluster to return the Infrastructure resource directly. But that helper wouldn't be all that useful, since it's just the Get() method of the client.

+1! I agree

> I'm not sure what you mean here. Can you give an example of what that might look like?

What I meant was - if we return nil or any other default value for non-openshift clusters, it would mean that we are trying to tell the user if the cluster supports HA or non HA mode when the cluster doesn't even expose such a metric. Would giving a specific error for this case be better?

Reply:
> > I'm not sure what you mean here. Can you give an example of what that might look like?
> 
> What I meant was - if we return nil or any other default value for non-openshift clusters, it would mean that we are trying to tell the user if the cluster supports HA or non HA mode when the cluster doesn't even expose such a metric. Wouldn't giving a specific error for this case be better?

I think I would need to see how we would expect a caller of this function to use it to answer. It sounds like the caller is going to have to do a bunch of work to check the return value and then decide how to handle error cases by default. Could we make it more helpful than that?

For example, could we trap all errors and return a default value provided by the caller if we can't retrieve a value from the Infrastructure CR, for any reason? That would at least let the caller avoid any extra logic when calling it. They could do something like

```go
topology := GetClusterTopology(HighlyAvailable)
```

without having to check for errors because we would promise to always return something.